src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/PathConverter.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities;
+
+using System.Collections.Generic;
+using System.Linq;
+using ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+using System.Collections.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.Utilities.Helpers.Interfaces;
+using System.IO;
+
+/// <summary>
+/// Converts paths in received and sent objects, to make testhost seem like it run a local test,
+/// while it was in fact running a test on a remote system, in a totally different path. This is for UWP which
+/// does testhost deployment.
+/// The modifications here rely on combination of side-effects, and actually replacing the values, because
+/// we cannot modify the properties on our public objects, and add setters.
+/// </summary>
+internal class PathConverter : IPathConverter
+{
+    // The path on this computer to which we deployed the test dll and test runner
+    private readonly string _deploymentPath = "";
+    // The path on the remote system where test dll was originally placed, and from which we
+    // copied it to this system. For vstest.console, which is on the other side of this, the names
+    // are inverted, it sends us their local path, and thinks about our local path as remote.
+    private readonly string _originalPath = "";
+
+    public PathConverter(string originalPath!!, string deploymentPath!!, IFileHelper fileHelper!!)
+    {
+        _originalPath = fileHelper.GetFullPath(originalPath).TrimEnd('\\').TrimEnd('/') + Path.DirectorySeparatorChar;
+        _deploymentPath = fileHelper.GetFullPath(deploymentPath).TrimEnd('\\').TrimEnd('/') + Path.DirectorySeparatorChar;
+    }
+
+    public string? UpdatePath(string? path, PathConversionDirection updateDirection)
+    {
+        if (path == null)
+            return path;
+
+        string find;
+        string replaceWith;
+        if (updateDirection == PathConversionDirection.Receive)
+        {
+            // Request is incoming, the path that is local to the sender (for us that is "remote" path)
+            // needs to be replaced with our path
+            find = _originalPath;
+            replaceWith = _deploymentPath;
+        }
+        else
+        {
+            find = _deploymentPath;
+            replaceWith = _originalPath;
+        }
+
+        var result = path?.Replace(find, replaceWith);
+        return result;
+    }
+
+    public IEnumerable<string?> UpdatePaths(IEnumerable<string?> paths!!, PathConversionDirection updateDirection)
+    {
+        return paths.Select(i => UpdatePath(i, updateDirection)).ToList();
```

Comment:
Just to match same style as other implementations:

```suggestion
        paths.ToList().ForEach(i => UpdatePath(i, updateDirection));
        return paths;
```

Reply:
This would not work, strings are immutable, so you would create a new string and output it to null, so the original string would not be updated.